fenicsadapter/fenicsadapter.py

```diff
@@ -259,501 +38,308 @@ def __init__(self, adapter_config_filename='precice-adapter-config.json',
         """
         self._fenics_dimensions = dimension
 
         if self._fenics_dimensions != self._dimensions:
-            logger.warning("fenics_dimension = {} and precice_dimension = {} do not match!".format(self._fenics_dimensions,
-                                                                                            self._dimensions))
-            if self._can_apply_2d_3d_coupling():
+            logger.warning(
+                "fenics_dimension = {} and precice_dimension = {} do not match!".format(self._fenics_dimensions,
+                                                                                        self._dimensions))
+            if can_apply_2d_3d_coupling(self._fenics_dimensions, self._dimensions):
                 logger.warning("2D-3D coupling will be applied. Z coordinates of all nodes will be set to zero.")
             else:
                 raise Exception("fenics_dimension = {}, precice_dimension = {}. "
                                 "No proper treatment for dimensional mismatch is implemented. Aborting!".format(
                     self._fenics_dimensions,
                     self._dimensions))
-        if dirichlet_boundary is not None:
-            self._Dirichlet_Boundary=dirichlet_boundary
 
-        self.set_coupling_mesh(mesh, coupling_subdomain)
-        self._set_read_field(read_field)
-        if write_field:
-            self._set_write_field(write_field)
-        self._precice_tau = self._interface.initialize()
+        self._set_coupling_mesh(mesh, coupling_subdomain)
+        precice_tau = self._interface.initialize()
+
+        # Adapter is initialized but first advance is yet to be called
+        self._first_advance_done = False
+
+        return precice_tau
+
+    def initialize_data(self, read_function, write_function, function_space):
+        """
+        Set initial conditions and boundary conditions
+        :param read_function: FEniCS function
+        :param write_function: FEniCS function
+        :param function_space: FEniCS space
+        :return:
+        """
+        self._read_function_type = determine_function_type(read_function)
+        self._read_function = read_function
+        self._function_space = function_space
 
-        if write_field:
-            if self._interface.is_action_required(precice.action_write_initial_data()):
-                self._write_block_data()
-                self._interface.mark_action_fulfilled(precice.action_write_initial_data())
+        if self._interface.is_action_required(action_write_initial_data()):
+            self.write(write_function)
+            self._interface.mark_action_fulfilled(action_write_initial_data())
 
         self._interface.initialize_data()
 
+        read_data = None
         if self._interface.is_read_data_available():
-            self._read_block_data()
+            read_data = self.read()
+
+        return read_data
+
+    def _set_coupling_mesh(self, mesh, subdomain):
```

Comment:
Do we need this function? Currently, it is only called once by `set_coupling_mesh`. Inlining this functions might decrease overall complexity of the class.